normandy/recipes/filters.py

```diff
@@ -816,6 +818,71 @@ def capabilities(self):
         return set()
 
 
+class JexlFilter(BaseFilter):
+    """
+    This filter allows the user to specify raw JEXL that will then be
+    included as a normal filter object.
+
+    It will combine with other filter objects like an other filter object,
+    that is it will be treated as a boolean expression and ANDed with it's
+    peers. The JEXL will by checked for syntactical validity. The expression
+    will be surrounded with parenthesis.
+
+    This filter should only be used when no other filter object can be used.
+
+    .. attribute:: type
+
+       ``jexl``
+
+    .. attribute:: expression
+       The expression to evaluate.
+
+       :example: ``2 + 2 >= 4``
+
+    .. attribute:: capabilities
+       An array of the capabilities required by the expression. May be empty
+       if the expression does not require any capabilities.
+
+       :example: ``["capabilities-v1"]``
+
+    .. attribute:: comment
+       A note about what this expression does. This field is not used
+       anywhere, but is present in the API to make it clearer what this
+       filter does.
+
+       :example: Only users that saw about:welcome.
+    """
+
+    type = "jexl"
+    expression = serializers.CharField()
+    capabilities = serializers.ListField(child=serializers.CharField(min_length=1), min_length=0)
+    comment = serializers.CharField(min_length=1)
+
+    def to_jexl(self):
+        built_expression = "(" + self.initial_data["expression"] + ")"
+        jexl = JEXL()
+
+        # Add mock transforms for validation. See
+        # https://mozilla.github.io/normandy/user/filters.html#transforms
+        # for a list of what transforms we expect to be available.
+        jexl.add_transform("date", lambda x: x)
+        jexl.add_transform("stableSample", lambda x: x)
+        jexl.add_transform("bucketSample", lambda x: x)
+        jexl.add_transform("preferenceValue", lambda x: x)
+        jexl.add_transform("preferenceIsUserSet", lambda x: x)
+        jexl.add_transform("preferenceExists", lambda x: x)
```

Comment:
i guess there's no way to grab the list of transforms and iterate to add these in?

Reply:
This is how we validate JEXL elsewhere as well, I just copy/pasted it. I couldn't really think of a better place to put it, and the other place is the `extra_filter_expression` evaluation, which we'll get rid of eventually. The server doesn't actually know what any of these transforms do, we just have to keep this in sync with the client.

A better thing would be to get information about the available transforms from the Normandy metadata that Firefox is publishing now, but maybe that's for another patch.

Reply:
> A better thing would be to get information about the available transforms from the Normandy metadata that Firefox is publishing now, but maybe that's for another patch.

+1 would love to not have to manually maintain this.